assembly/nfa/characters.ts

```diff
@@ -1,32 +1,34 @@
-export function isDigit(code: i32): bool {
-  return code >= 48 && code <= 57;
+export function isDigit(code: u32): bool {
+  return code - 48 < 10;
 }
 
-export function isLowercaseAlpha(code: i32): bool {
-  return code >= 97 && code <= 122;
+export function isLowercaseAlpha(code: u32): bool {
+  return code - 97 < 26;
 }
 
-export function isUppercaseAlpha(code: i32): bool {
-  return code >= 65 && code <= 90;
+export function isUppercaseAlpha(code: u32): bool {
+  return code - 65 < 26;
 }
 
-export function isUnderscore(code: i32): bool {
+export function isUnderscore(code: u32): bool {
   return code == 95;
 }
 
-export function isWhitespace(code: i32): bool {
-  return (
-    (code >= 9 && code <= 13) ||
-    code == 32 ||
-    code == 0x00a0 ||
-    code == 0x1680 ||
-    code == 0x2000 ||
-    code == 0x200a ||
-    code == 0x2028 ||
-    code == 0x2029 ||
-    code == 0x202f ||
-    code == 0x205f ||
-    code == 0x3000 ||
-    code == 0xfeff
-  );
+export function isWhitespace(code: u32): bool {
+  if (code < 0x1680) { // < <LS> (1)
+    // <SP>, <TAB>, <LF>, <VT>, <FF>, <CR> and <NBSP>
+    // @ts-ignore: cast
+    return ((code | 0x80) == 0xA0) | (code - 0x09 <= 0x0D - 0x09);
+  }
+  if (code - 0x2000 <= 0x200A - 0x2000) return true;
+  switch (code) {
+    case 0x1680: // <LS> (1)
+    case 0x2028: // <LS> (2)
+    case 0x2029: // <PS>
+    case 0x202F: // <NNS>
+    case 0x205F: // <MMSP>
+    case 0x3000: // <IS>
+    case 0xFEFF: return true; // <ZWNBSP>
+  }
+  return false;
```

Comment:
note all this code is not properly transpilable to JS by `tsc` due to based on unsigned integer operations (code has `u32`). If you want more portable code you probably to follow two different approaches:
1) Add JS-style casts after every relational ops. like: `((code - 0x2000) >>> 0) <= ((0x200A - 0x2000) >>> 0)`. Or you could use AS macro:
```ts
if (ASC_TARGET == 1) {
  // AS code with unsigned ops
} else {
  // JS compatible code which use `>>> 0` or different implementation
}
```

Reply:
Thanks for pointing this out. I do make extensive use of the ability to run this code as TS=>JS as well as AS=>wasm - I find the debugging experience much easier with TS/JS!